File: akf_hrms/akf_hrms/report/turn_over_report/turn_over_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

// Mubashir Bashir

frappe.query_reports["Turn over report"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"duration",
			"label": __("Duration"),
			"fieldtype": "Select",
			"options": ["Monthly", "Biannually", "Annually"],
			"reqd": 1,
			on_change: function() {
				var start_date = frappe.datetime.month_start(frappe.datetime.nowdate()); 
				var end_date = frappe.datetime.month_end(frappe.datetime.nowdate());
				var duration_value = frappe.query_report.get_filter_value('duration')
				if (duration_value == "Monthly"){
					frappe.query_report.set_filter_value('from_date',start_date);
				}
				else if (duration_value == "Biannually"){
					var sub_months = frappe.datetime.add_months(start_date, -5); 
					frappe.query_report.set_filter_value('from_date',sub_months);
				}
				else{
					var sub_months = frappe.datetime.add_months(start_date, -11); 
					frappe.query_report.set_filter_value('from_date',sub_months);
				}
				frappe.query_report.set_filter_value('to_date',end_date);
			}
		},
		{
			"fieldname":"department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department"
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"hidden": 1
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"hidden": 1
		},
		{
			"fieldname":"status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": ["","Active", "Left"]
		}
	]
}

File: akf_hrms/akf_hrms/report/turn_over_report/turn_over_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

# Mubashir Bashir

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
	columns, data = [], []

	columns = get_columns(filters)

	data = get_data(filters)

	return columns, data


def get_columns(filters): 
	return [
		_("Department")+ ":Data:250",_("Total Employees")+ ":Data:250"
	]

def get_data(filters):
	total_emp_count = []
	conditions, filters = get_conditions(filters)

	if filters.get("department"):
		emp_info = frappe.db.sql("""select count(*) from `tabEmployee` where docstatus < 2 %s""" % conditions, filters)
		total_emp_count.append([filters.get("department"),emp_info[0][0]])
	else:
		total_dep = frappe.db.sql("""select name from `tabDepartment` where docstatus < 2 """)
		for x in total_dep:
			emp_info = frappe.db.sql("""select count(*) from `tabEmployee` where department = '%s' %s """ % (x[0], conditions), filters )
			total_emp_count.append([x[0],emp_info[0][0]])
	return total_emp_count

def get_conditions(filters):
	conditions = ""
	if filters.get("company"): conditions += " and company = %(company)s"
	if filters.get("department"): conditions += " and department = %(department)s"
	if filters.get("status") == 'Active':
		conditions += " and date_of_joining between %(from_date)s and %(to_date)s and status = 'Active'"
	if filters.get("status") == 'Left':
		conditions += " and relieving_date between %(from_date)s and %(to_date)s and status = 'Left'"
	if filters.get("status") != 'Active' and filters.get("status") != 'Left':
		conditions += " and ((date_of_joining between %(from_date)s and %(to_date)s) or (relieving_date between %(from_date)s and %(to_date)s))"
	return conditions, filters

